Create a generalized uniform superposition state gate

prag16 opened the pull request:
Creates a generalized uniform superposition state, $$\lvert \Psi \rangle = \frac{1}{\sqrt{M}} \sum_{j=0}^{M-1} \ket{j}$$ (where $1 < M \leq 2^n$), using $n$ qubits, according to the Shukla-Vedula algorithm [SV24]. Note that when $M$ is of the form $2^n$, creating uniform superposition states is trivial (just apply $n$ Hadamard gates on $\ket{0}^{\otimes n}$). However, in the general case, when $M$ is not of this form, creating uniform superposition states is non-trivial. The Shukla-Vedula algorithm [SV24] offers an efficient approach for the creation of a generalized uniform superposition state $ \lvert \Psi \rangle$ (as defined above) requiring only $O(\log_2 (M))$ qubits and $O(\log_2 (M))$ gates. This provides an exponential improvement (in the context of reduced resources and complexity) over other approaches in the literature.

Args:
        M (integer): 
            A positive integer $M (> 1)$ representing the number of computational basis states with an amplitude of $\frac{1}{\sqrt{M}}$ in the uniform superposition state $$\lvert \Psi \rangle = \frac{1}{\sqrt{M}} \sum_{j=0}^{M-1} \ket{j}.$$ Note that the remaining $(2^n - M)$ computational basis states have zero amplitudes. Here $M$ need not be an integer power of $2$.
            
        n (integer): 
            A positive integer representing the number of qubits used.

    Returns:
        A quantum circuit that creates the uniform superposition state: $$\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1} \ket{j}$$. 

References:
[SV24] A. Shukla and P. Vedula, “An efficient quantum algorithm for the preparation of uniform quantum superposition states,” Quantum Information Processing, 23(38), pp. 1-32 (2024).

prag16 commented on cirq-core/cirq/ops/generalized_uniform_superposition_gate.py:
```diff
@@ -0,0 +1,92 @@
+    """
+
+    if (num_qubits < np.log2(M)):
+        print('Error Message: Not enough qubits! Try increasing num_qubits ..')
```
Changed as suggested.

prag16 commented on cirq-core/cirq/ops/generalized_uniform_superposition_gate.py:
```diff
@@ -0,0 +1,92 @@
+        print('Error Message: Not enough qubits! Try increasing num_qubits ..')
+        return
+
+    qreg = cirq.LineQubit.range(num_qubits)
```
Thank you for this suggestion. Following this feedback, I implemented a gate-based approach.

prag16 commented on cirq-core/cirq/ops/generalized_uniform_superposition_gate.py:
```diff
@@ -0,0 +1,92 @@
+    qreg.reverse()
+    qcircuit = cirq.Circuit()
+
+    # Delete the line below
```
Done.

prag16 commented on cirq-core/cirq/ops/generalized_uniform_superposition_gate.py:
```diff
@@ -0,0 +1,92 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import cirq
```
Thank you for the feedback. Now, only specific sub-packages are being imported.

prag16 commented on cirq-core/cirq/ops/generalized_uniform_superposition_gate.py:
```diff
@@ -0,0 +1,92 @@
+import cirq
+import numpy as np
+
+def generalized_uniform_superposition_cirqx(M, num_qubits):
```
Based on your other comment, now I've implemented a gate-based code. I have taken care to provide the correct type informaiton in the new code.

prag16 commented on cirq-core/cirq/ops/generalized_uniform_superposition_gate.py:
```diff
@@ -0,0 +1,92 @@
+import cirq
+import numpy as np
+
+def generalized_uniform_superposition_cirqx(M, num_qubits):
```
Thanks for the comments. The suffix was removed in the updated version.

prag16 replied:
I would appreciate it if you could please review the changes and provide your feedback/approval at your earliest convenience.

prag16 commented on cirq-core/cirq/ops/generalized_uniform_superposition_gate.py:
```diff
@@ -0,0 +1,92 @@
+    for i in L[1:k]:
+        qcircuit.append(cirq.X(qreg[i]))
+
+    Mcurrent = 2**(L[0])
```
Thanks for the feedback. I followed the notation used in the referenced paper. It would be helpful to keep it this way and maintain the same variable names. I'll stick to the suggested naming convention in my future submissions.

prag16 commented on cirq-core/cirq/ops/generalized_uniform_superposition_gate.py:
```diff
@@ -0,0 +1,92 @@
+            
+        Mcurrent = Mcurrent + 2**(L[m])
+
+    qreg.reverse()
```
Removed this form the updated code.

prag16 commented on cirq-core/cirq/ops/generalized_uniform_superposition_gate_test.py:
```diff
@@ -0,0 +1,40 @@
+import cirq
+import numpy as np
+
+def check_uniform_superposition_error(M, n):
```
Thank you for the feedback. Suggested changes were made.

prag16 commented:
@vtomole @dstrain115  - The code appears to be working fine on my end. Could you please review it and provide your feedback?

tanujkhattar commented:
What's the use case for this? Why are we adding this to Cirq? 

>  However, in the general case, when is not of this form, creating uniform superposition states is non-trivial. 

Uniform superposition states can be prepared using a single round of amplitude amplification using a single ancilla qubit. See https://github.com/quantumlib/Qualtran/blob/b9c5b5bd33d17f35fe027ff5362dd0b2f6887048/qualtran/bloqs/state_preparation/prepare_uniform_superposition.py#L29 for more details. 

Can you highlight what's the difference of your approach against previously known results like Fig 12 of https://arxiv.org/pdf/1805.03662.pdf which is implemented in the Qualtran link I've shared above? Also, I think Qualtran might be a better repository for algorithmic primitives like this. 

@dstrain115 Is this needed in Cirq for a specific use case I'm not aware of?

prag16 commented:
@tanujkhattar Thanks for your comments. Please note the following points about the approach suggested in Fig 12 of https://arxiv.org/pdf/1805.03662.pdf vs the approach used in the code in this PR:
(1) It requires 2⌈log L⌉ + O(1) qubits, which is double the number of qubits needed by the submitted code in this PR 
(2) The T count given in the arxiv reference is 2k + 10 log L + O(log( 1/ϵ )) with a dependence on O(log( 1/ϵ )),
The submitted code in this PR needs only O(log L) gates without any dependence on O(log( 1/ϵ )) term.
(3) Lastly, but most importantly the submitted code is deterministic and 100% succeeds, whereas the code in the arxiv reference is probabilistic (it uses amplitude amplification). Theoretical computation shows that the probability of success can be significantly less (less than 12%) in many instances when one uses the probabilistic algorithm given in the arxiv algorithm. 

Considering the above points, it seems the submitted code provides a simpler (just a few lines of code), cleaner, and deterministic approach for the preparation of uniform superposition states as compared to the probabilistic approach given in the arxiv reference.

Next coming to - What's the use case for this? Why are we adding this to Cirq? I guess creating uniform superposition states is an important algorithm - it is the first step in many important quantum algorithms - including the case when the number of uniform superposition states N is not of the form of 2^n, which will be useful for a variety of applications. I am open to your comments and suggestions.

tanujkhattar commented:
Can you share a quirk link of the circuits produced by your algorithm and compare it with the circuits I've linked in the reference for a better understanding and comparison? See [this link for what the circuit looks like for the linked reference](  
https://algassert.com/quirk#circuit=%7B%22cols%22%3A%5B%5B1%2C1%2C1%2C1%2C1%2C1%2C%7B%22id%22%3A%22setB%22%2C%22arg%22%3A25%7D%5D%2C%5B%5D%2C%5B%22H%22%2C%22H%22%2C%22H%22%2C%22H%22%2C%22H%22%2C%22H%22%2C%22H%22%5D%2C%5B1%2C1%2C%22inputA5%22%2C1%2C1%2C1%2C1%2C%22%5EA%3CB%22%5D%2C%5B1%2C1%2C1%2C1%2C1%2C1%2C1%2C%7B%22id%22%3A%22Rzft%22%2C%22arg%22%3A%22acos(1-16%2F25)%22%7D%5D%2C%5B%5D%2C%5B%5D%2C%5B1%2C1%2C%22inputA5%22%2C1%2C1%2C1%2C1%2C%22%5EA%3CB%22%5D%2C%5B1%2C1%2C%22%E2%8A%95%22%2C%22%E2%8A%95%22%2C%22%E2%8A%95%22%2C%22%E2%8A%95%22%2C%22%E2%8A%95%22%2C%22X%22%5D%2C%5B1%2C1%2C1%2C1%2C1%2C1%2C1%2C%7B%22id%22%3A%22Rzft%22%2C%22arg%22%3A%22acos(1-16%2F25)%22%7D%5D%2C%5B%5D%2C%5B%5D%2C%5B1%2C1%2C%22%E2%8A%95%22%2C%22%E2%8A%95%22%2C%22%E2%8A%95%22%2C%22%E2%8A%95%22%2C%22%E2%8A%95%22%2C%22X%22%5D%2C%5B%22Chance7%22%5D%2C%5B%22Amps7%22%5D%5D%7D)

> (1) The approach in the arxiv reference requires 2⌈log L⌉ + O(1) qubits (in the interesting case when L is odd), which is double the number of qubits needed by the submitted code in this PR

That's correct. The additional `logL` qubits are used in the `LessThanGate` oracle to minimize the T-count. \

> (2) The T count given in the arxiv reference is 2k + 10 log L + O(log( 1/ϵ )) with a dependence on O(log( 1/ϵ )),
The submitted code in this PR needs only O(log L) gates without any dependence on O(log( 1/ϵ )) term.

The dependence on `O(log( 1/ϵ ))` comes from how closely you approximate a single `Ry(theta)` rotation gate and is relevant in the fault tolerant regime. Your implementation also has this dependence since you use multiple small angle rotation of the form `cirq.ry(theta)` and controlled versions of it thereof. 

> (3) Lastly, but most importantly the submitted code is deterministic and 100% succeeds, whereas the code in the arxiv 
reference is probabilistic (it uses amplitude amplification). Theoretical computation shows that the probability of success can be significantly less (less than 12%) in many instances when one uses the probabilistic algorithm given in the arxiv reference.

The code in the linked reference is also deterministic. When you know the initial overlap, you can perform amplitude amplification to deterministically end on the good state. See the quirk link above to understand the circuit in the linked reference better.

prag16 commented:
@tanujkhattar  Thanks for sharing the link. As you suggested, I am sharing the links to quirk circuits for the cases [N=100](https://algassert.com/quirk#circuit=%7B%22cols%22%3A%5B%5B%22H%22%2C%22H%22%2C1%2C1%2C1%2C%22X%22%2C%22X%22%5D%2C%5B1%2C1%2C1%2C1%2C1%2C%7B%22id%22%3A%22Ryft%22%2C%22arg%22%3A%22-2.74%22%7D%5D%2C%5B%5D%2C%5B1%2C1%2C%22H%22%2C%22H%22%2C%22H%22%2C%22%E2%97%A6%22%2C%7B%22id%22%3A%22Ryft%22%2C%22arg%22%3A%22-1.91%22%7D%5D%2C%5B%5D%2C%5B1%2C1%2C1%2C1%2C1%2C%22H%22%2C%22%E2%97%A6%22%5D%5D%7D), [N=130](https://algassert.com/quirk#circuit={%22cols%22:[[%22H%22,1,1,1,1,1,1,%22X%22],[1,1,1,1,1,1,1,{%22id%22:%22Ryft%22,%22arg%22:%22-2.89%22}],[],[1,%22H%22,%22H%22,%22H%22,%22H%22,%22H%22,%22H%22,%22%E2%97%A6%22]]}), and [N=127](https://algassert.com/quirk#circuit={%22cols%22:[[1,%22X%22,%22X%22,%22X%22,%22X%22,%22X%22,%22X%22],[1,{%22id%22:%22Ryft%22,%22arg%22:%22-2.96%22}],[],[%22H%22,%22%E2%97%A6%22,{%22id%22:%22Ryft%22,%22arg%22:%22-2.89%22}],[],[1,%22H%22,%22%E2%97%A6%22,{%22id%22:%22Ryft%22,%22arg%22:%22-2.78%22}],[],[1,1,%22H%22,%22%E2%97%A6%22,{%22id%22:%22Ryft%22,%22arg%22:%22-2.62%22}],[],[1,1,1,%22H%22,%22%E2%97%A6%22,{%22id%22:%22Ryft%22,%22arg%22:%22-2.37%22}],[],[1,1,1,1,%22H%22,%22%E2%97%A6%22,{%22id%22:%22Ryft%22,%22arg%22:%22-1.91%22}],[],[1,1,1,1,1,%22H%22,%22%E2%97%A6%22]]}) based on the approach used in the submitted PR.  
I think the above circuit for N=100 is much simpler compared to the circuit in the link that you provided for the same case. In other cases also, the provided circuits are very simple and efficient. The above circuits use fewer number qubits (which is consistent with what we discussed earlier about the arxiv reference requiring 2⌈log L⌉ + O(1) qubits  which is double the number of qubits needed by the submitted code in this PR). The number of gates and the circuit depth in the submitted PR are considerably lower than the arxiv reference that you shared (which is consitent with the T count given in the arxiv reference 2k + 10 log L + O(log( 1/ϵ ))). The use of multi-controlled rotation gates and comparator circuits increases the circuit gate complexity for the approach given in the arxiv reference, whereas only at most single-controlled gates are needed in the approach that I shared.

Finally, typically algorithms involving amplitude amplification are probabilistic, unless one can show that the probability of obtaining the bad states is 0 after the amplitude amplification. This proof or the expression for the probability of success was not provided in the reference that you shared. However, The authors of the reference you shared have discussed the probabilistic nature of a later version of their algorithm (for preparation of uniform superposition states) in a later paper and provided an expression for the probability of success. Please refer to Eq. 205 and Eq. 206 and Figure 10 in [Compilation of Fault-Tolerant Quantum Heuristics for Combinatorial Optimization](https://arxiv.org/pdf/2007.07391.pdf)

Even if we keep aside the discussion on the probabilistic nature of algorithm of the arxiv reference, considering the discussion above and the quirk circuits provided, it is clear that the submitted code in the PR provides a cleaner approach with fewer resources for the preparation of the uniform superposition states.  

Overall, I think there is merit in the potential inclusion of the submitted code in Cirq  framework.

NoureldinYosri commented:
I have added the `triage/discuss` label so that we discuss this PR in the next cirq-sync meeting

@prag16 the next cirq-sync will be on May 8th. you can either join the [cirq-dev group](https://groups.google.com/g/cirq-dev) to get an automatic invitation or I can send you an invitation to the meeting (in that case which email should I use).

prag16 commented:
@NoureldinYosri Thank you for the invitation. I have joined the cirq-dev group and am looking forward to attending the next cirq-sync on May 8th.

anurudhp commented on cirq-core/cirq/ops/generalized_uniform_superposition_gate_test.py:
```diff
@@ -0,0 +1,46 @@
+    )
+
+
+def _test1_check_uniform(): 
```
```suggestion
def test1_check_uniform(): 
```
please fix this: unit tests should start with `test_`

anurudhp replied:
On renaming this, the test does not pass, please fix that.

prag16 replied:
Thank you for the suggestion. I've updated the code, including the import to UniformSuperpositionGate in the cirq-core/cirq/__init__.py and cirq-core/cirq/ops/__init__.py files. This was done to address the error encountered during unit testing. Also, I've renamed certain variables and files to simplify them and align with the pylint coding standard.  Hopefully all the tests should now pass successfully.

NoureldinYosri commented on cirq-core/cirq/ops/uniform_superposition_gate.py:
```diff
@@ -0,0 +1,145 @@
+        qreg = list(qubits)
+
+        if (self._m_value & (self._m_value - 1)) == 0:  # if m_value is an integer power of 2
+            m = int(math.log2(self._m_value))
```
```suggestion
m = self._m_value.bit_length() - 1
yield H.on_each(qreg[:m])
```

prag16 replied:
@NoureldinYosri - Thanks for the suggestion. The code changed accordingly.

NoureldinYosri commented on cirq-core/cirq/ops/uniform_superposition_gate_test.py:
```diff
@@ -0,0 +1,47 @@
+import cirq
+
+
+def _assert_generated_unitary_is_uniform(m_value: int, n: int) -> None:
```
prefer to use parameterized tests
 
```suggestion
@pytest.mark.parameterize(['m', 'n'], TEST_VALUES)
def test_generated_unitary_is_uniform(m: int, n: int) -> None:
```

prag16 replied:
I have added parametrized tests for the test_uniform_superposition_gate() with low_value and max_values as test parameters.

NoureldinYosri replied:
that doesn't serve the purpose of parameterization since it's still the same test as before. the point of parameterization is to isolate each test case so that when one of them fails we would know exactly which one.

NoureldinYosri replied:
```suggestion
@pytest.mark.parameterize(['m', 'n'], 
   [[m, n] for n in range(3, 7) for m in np.random.randint(1, 1 << n, size=3)
)
def test_generated_unitary_is_uniform(m: int, n: int) -> None:
```

NoureldinYosri reviewed (changes requested):
please prefer to parameterize tests + a few nits

prag16 commented on cirq-core/cirq/ops/uniform_superposition_gate.py:
```diff
@@ -0,0 +1,145 @@
+                yield H(qreg[i])
+            return
+
+        n_value = [int(x) for x in list(np.binary_repr(self._m_value))][::-1]
```
@NoureldinYosri - Thank you for the suggestion. I have made the suggested changes to the code.

prag16 commented on cirq-core/cirq/ops/uniform_superposition_gate.py:
```diff
@@ -0,0 +1,145 @@
+
+        qreg.reverse()
+
+        for i in l_value[1:k]:
```
Thank you. I have made the suggested changes.

NoureldinYosri commented on cirq-core/cirq/ops/uniform_superposition_gate.py:
```diff
@@ -0,0 +1,144 @@
+
+            m_current = m_current + 2 ** (l_value[m])
+
+    def num_qubits(self) -> int:
```
nit: no need for docstring here or in  __repr__

prag16 replied:
docstring removed.

NoureldinYosri commented on cirq-core/cirq/ops/uniform_superposition_gate.py:
```diff
@@ -0,0 +1,144 @@
+        """
+        return self._num_qubits
+
+    def __repr__(self) -> str:
```
does any test fail with the default __repr__? if not please remove this function

prag16 replied:
The function was removed.

NoureldinYosri commented on cirq-core/cirq/ops/uniform_superposition_gate_test.py:
```diff
@@ -0,0 +1,47 @@
+    )
+
+@pytest.mark.parametrize('low_value, max_value', [(3,17),(17,29),(30,89),(89,127),(128,257)])
+def test_uniform_superposition_gate(low_value,max_value):
```
please remove this function

prag16 replied:
Thank you. I have fixed the code following your suggestion.

NoureldinYosri commented on cirq-core/cirq/ops/uniform_superposition_gate_test.py:
```diff
@@ -17,7 +17,7 @@
 
 @pytest.mark.parameterize(['m', 'n'],[[m, n] for n in range(3, 7)
-                                      for m in np.random.randint(1, 1 << n, size=3)])
+                                      for m in np.random.randint(2, 1 << n, size=3)])
```
please undo this chage, the gate should work even with `m=1`

prag16 replied:
This change was reverted back as suggested. Since at least two states are needed to create a superposition, for m=1, an appropriate value error is raised. The test case should now also work for m =1.

NoureldinYosri replied:
`m=1` means $\ket{0 \cdots 0}$ (i.e. do nothing). so 
```py3
if m == 1: return
```

prag16 replied:
Thanks for the suggestion. I have updated the code accordingly. I have also included several test cases to ensure comprehensive coverage of all code lines.

prag16 replied:
@NoureldinYosri - Could you please check if the tests are passing now?

prag16 replied:
@NoureldinYosri - Thanks for running the test, I've now fixed the remaining minor issues. Hopefully, this should now successfully pass.  Could you please run the tests again?

anurudhp commented on cirq-core/cirq/ops/uniform_superposition_gate.py:
```diff
@@ -0,0 +1,127 @@
+
+        for m in range(1, len(l_value) - 1):
+            theta = -2 * np.arccos(np.sqrt(2 ** l_value[m] / (self._m_value - m_current)))
+            yield controlled_gate.ControlledGate(ry(theta), control_values=[False])(
+                qreg[l_value[m]], qreg[l_value[m + 1]]
+            )
```
```suggestion
            yield ry(theta).on(qreg[l_value[m + 1]]).controlled_by(
                qreg[l_value[m]], control_values=[0]
            )
```

prag16 replied:
Thanks. Suggested changes were made

anurudhp commented on cirq-core/cirq/ops/uniform_superposition_gate.py:
```diff
@@ -0,0 +1,127 @@
+                qreg[l_value[m]], qreg[l_value[m + 1]]
+            )
+            for i in range(l_value[m], l_value[m + 1]):
+                yield controlled_gate.ControlledGate(H, control_values=[False])(
+                    qreg[l_value[m + 1]], qreg[i]
+                )
```
```suggestion
                yield H.on(qreg[i]).controlled_by(qreg[l_value[m + 1]], control_values=[0])
```

prag16 replied:
@anurudhp Thanks for your input. Suggested changes were made

anurudhp commented on cirq-core/cirq/ops/uniform_superposition_gate_test.py:
```diff
@@ -0,0 +1,62 @@
+            gate = cirq.UniformSuperpositionGate(m, n)
+    else:
+        gate = cirq.UniformSuperpositionGate(m, n)
+        qregx = cirq.LineQubit.range(n)
+        qcircuit = cirq.Circuit(gate.on(*qregx))
+
+        unitary_matrix1 = np.real(qcircuit.unitary())
```
```suggestion
        matrix = np.real(cirq.unitary(gate))
```

prag16 replied:
@anurudhp Thank you for your suggestion. I made the suggested changes. Can you please check if the tests are passing now?

prag16 commented:
@NoureldinYosri and @anurudhp  - Thank you very much for all your helpful suggestions. Now, all tests are passing except for one test (failing due to a JSON-related issue, which I have now fixed). I am confident that the test will pass now. Could you please run the test again?

prag16 commented on cirq-core/cirq/ops/uniform_superposition_gate_test.py:
```diff
@@ -0,0 +1,62 @@
+    [[int(m), n] for n in range(3, 7) for m in np.random.randint(1, 1 << n, size=3)]
+    + [(1, 1), (-2, 1), (-3.1, 2), (6, -4)],
+)
+def test_generated_unitary_is_uniform(m: int, n: int) -> None:
```
Thanks for the suggestions. I have split the test into two separate tests as suggested.

prag16 commented on cirq-core/cirq/ops/uniform_superposition_gate_test.py:
```diff
@@ -0,0 +1,59 @@
+            gate = cirq.UniformSuperpositionGate(m, n)
+    else:
+        gate = cirq.UniformSuperpositionGate(m, n)
+        matrix = np.real(cirq.unitary(gate))
```
Thank you. I'm now using np.array instead of np.real to take into account this.

prag16 commented on cirq-core/cirq/ops/uniform_superposition_gate.py:
```diff
@@ -0,0 +1,125 @@
+
+
+class UniformSuperpositionGate(raw_types.Gate):
+    r"""Creates a uniform superposition state, $\frac{1}{\sqrt{M}}\sum_{j=0}^{M-1}\ket{j}$
```
Suggested changes have been made.

prag16 commented on cirq-core/cirq/ops/uniform_superposition_gate_test.py:
```diff
@@ -0,0 +1,59 @@
+    elif n < np.log2(m):
+        with pytest.raises(
+            ValueError,
+            match='num_qubits must be an integer greater than or equal to log2\\(m_value\\).',
```
Unfortunately, the regex is not correctly parsing (m_value) without using \\(m_value\\) and the test was failing locally with (m_value). Therefore, I kept \\(m_value\\) for now.

prag16 commented on cirq-core/cirq/ops/uniform_superposition_gate.py:
```diff
@@ -0,0 +1,125 @@
+    generalized uniform superposition state of the form,
+    $\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j} $, requiring only $O(\log_2 (M))$ qubits and
+    $O(\log_2 (M))$ gates. This provides an exponential improvement (in the context of reduced
+    resources and complexity) over other approaches in the literature.
```
Done.

prag16 commented on cirq-core/cirq/ops/uniform_superposition_gate.py:
```diff
@@ -0,0 +1,125 @@
+
+    References:
+        [SV24]
+            A. Shukla and P. Vedula, "An efficient quantum algorithm for preparation of uniform
```
Done.

prag16 commented on cirq-core/cirq/ops/uniform_superposition_gate.py:
```diff
@@ -0,0 +1,125 @@
+        """Initializes UniformSuperpositionGate.
+
+        Args:
+            m_value (int): The number of computational basis states with amplitude 1/sqrt(M).
```
Done.

prag16 commented on cirq-core/cirq/ops/uniform_superposition_gate.py:
```diff
@@ -0,0 +1,125 @@
+
+        Args:
+            m_value (int): The number of computational basis states with amplitude 1/sqrt(M).
+            num_qubits (int): The number of qubits used.
```
Done.

prag16 commented on cirq-core/cirq/ops/uniform_superposition_gate.py:
```diff
@@ -0,0 +1,125 @@
+        """
+        if not (isinstance(m_value, int) and (m_value > 0)):
+            raise ValueError('m_value must be a positive integer.')
+        if not (isinstance(num_qubits, int) and (num_qubits >= math.log2(m_value))):
```
Thank you for the suggestion. I have made this change.

prag16 replied:
In fact, m_value.bit_length() is not the same as math.log2(m_value) when m_value is of the form 2^k for some positive integer k. I have fixed this error now.

prag16 commented on cirq-core/cirq/ops/uniform_superposition_gate.py:
```diff
@@ -0,0 +1,125 @@
+    def _decompose_(self, qubits: Sequence['cirq.Qid']) -> 'cirq.OP_TREE':
+        """Decomposes the gate into a sequence of standard gates.
+
+        Args:
```
Done.

prag16 commented:
@NoureldinYosri and @anurudhp: Thank you very much for all your helpful suggestions. I have made the suggested changes. 

I am confident that the tests will pass now. Could you please run the tests to check if they are passing now?

NoureldinYosri commented on cirq-core/cirq/ops/uniform_superposition_gate.py:
```diff
@@ -0,0 +1,104 @@
+class UniformSuperpositionGate(raw_types.Gate):
+    r"""Creates a uniform superposition state on the states $[0, M)$
+    The gate creates the state $\frac{1}{\sqrt{M}}\sum_{j=0}^{M-1}\ket{j}$
+    (where 1<= M <= 2^n), using n qubits, according to the Shukla-Vedula algorithm [SV24].
```
```suggestion
    (where $1\leq M \leq 2^n$), using n qubits, according to the Shukla-Vedula algorithm [SV24].
```

prag16 replied:
@NoureldinYosri - Thank you for the suggestion. I made the suggested changes. The format check is passing locally for me now. This should pass now in CI as well.

NoureldinYosri reviewed (commented):
one more nit + fix ci

you can fix the format by running `./check/format-incremental --apply`. you can also run the ci locally to make sure they pass https://github.com/quantumlib/Cirq/blob/bf67f2973adf6c0c56fd32a19d1f161449d8ee7b/docs/dev/development.md#continuous-integration-and-local-testing

NoureldinYosri reviewed (commented):
@prag16 Hi gates defined in`cirq.ops` have to provide certain functionalities so please don't turnoff any tests. if you don't want to implement all the functionalities required for a cirq.Gate then you can move the gate to `cirq-core/cirq/contrib` where we accept nonproduction contributions

prag16 commented:
@NoureldinYosri -  I turned off the json-related tests temporarily to ensure that all the other tests are passing first (and we are pretty close to that). Next, after including the json-related functionality, I'll turn these tests back on and request your review.

prag16 commented:
@NoureldinYosri - I implemented the json-related functionality and turned on all the tests. All the checks are passing at my end. Can you please run the ci tests at your end to confirm?

NoureldinYosri reviewed (commented):
LGTM%fix failing CI

prag16 commented:
@NoureldinYosri - Thanks for reviewing the changes. Now, I've corrected the formatting and added more tests to fix CI coverage check errors.  Hopefully, the CI tests should pass now.

prag16 commented:
@NoureldinYosri - I have taken care of the last remaining coverage error and also fixed the format error. The CI checks are passing locally for me, Hopefully, the checks will pass at your end as well.

NoureldinYosri reviewed (approved):
@prag16 thanks for your contribution, great work

prag16 commented:
@NoureldinYosri - Thank you very much for reviewing the code and for your helpful suggestions! I really appreciate it!
